### ISSUE TITLE
Sharded ttnn.matmul has bad output when program cache is enabled and output mem config is set to interleaved

### ISSUE THREAD

**Issue by mtairum** [tenstorrent/tt-metal#8512]
Sharded `ttnn.matmul` has a bad output when **program cache is enabled** and the **output memory config is set to interleaved**. Tested on single wormhole chip on latest main.

The issue can be seen when running multiple loops (2 is enough) of the same matmul with the same inputs.

In the reproduction code below, the `matmul` sets its memory config to interleaved in L1 `ttnn.L1_MEMORY_CONFIG`. (error also appears on DRAM). Both the PCC comparison and `torch.allclose()` will fail.

If however we change the matmul output to be sharded, using `out_config` instead and then explicitly interleave the tensor with `attn_1BQP = ttnn.to_memory_config(attn_1BQP, memory_config=ttnn.L1_MEMORY_CONFIG)` the test will now pass. This is the expected output.

This behaviour only happens when program cache is enabled and more than one loop is done.

## Reproduction code

Run the code below with pytest to make use of the test fixtures.

```
import torch
import ttnn
from models.utility_functions import comp_pcc

def test_sharded_matmul(device, use_program_cache):
    q_config = ttnn.create_sharded_memory_config(
            shape=(32, 128),
            core_grid=ttnn.CoreGrid(y=4, x=8),
            strategy=ttnn.ShardStrategy.HEIGHT,
            orientation=ttnn.ShardOrientation.ROW_MAJOR,
            use_height_and_width_as_shard_shape=True,
            )
    k_config = ttnn.create_sharded_memory_config(
            shape=(128, 32),
            core_grid=ttnn.CoreGrid(y=4, x=8),
            strategy=ttnn.ShardStrategy.HEIGHT,
            orientation=ttnn.ShardOrientation.ROW_MAJOR,
            use_height_and_width_as_shard_shape=True,
            )
    out_config = ttnn.create_sharded_memory_config(
        shape=(32, 32),
        core_grid=ttnn.CoreGrid(y=4, x=8),
        strategy=ttnn.ShardStrategy.HEIGHT,
        orientation=ttnn.ShardOrientation.ROW_MAJOR,
        use_height_and_width_as_shard_shape=True,
    )

    q_heads_1BQD= ttnn.from_torch(torch.randn(1, 32, 32, 128), device=device, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat8_b, memory_config=q_config)
    keys_1BDP= ttnn.from_torch(torch.randn(1, 32, 128, 32), device=device, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat8_b, memory_config=k_config)
    results = []
    for _ in range(2):
        attn_1BQP = ttnn.matmul(
            q_heads_1BQD,
            keys_1BDP,
            core_grid=ttnn.CoreGrid(y=4, x=8),
            dtype=ttnn.bfloat16,
            # memory_config = out_config,
            memory_config = ttnn.L1_MEMORY_CONFIG,
        )
        results.append(ttnn.to_torch(attn_1BQP))

        # explicitly convert sharded to interleave again
        # attn_1BQP = ttnn.to_memory_config(attn_1BQP, memory_config=ttnn.L1_MEMORY_CONFIG)

    
    passing, _ = comp_pcc(results[0], results[1])
    assert passing, "PCC fail"
    assert torch.allclose(results[0], results[1]), "torch.allclose() fail"

```

**Comment by mtairum** [comment]
Added P1 label since it looks like an unintended bug.
However the workaround is also given in the ticket, so this can be bumped down to P2 if needed.

**Comment by prajaramanTT** [comment]
@mtairum Is this still an open issue ? If not, can you please mark this closed ? Thanks.

**Comment by mtairum** [comment]
Tested on latest main. This is no longer an issue. Closing.
